### PR TITLE
allow pack task to get properties for token substitution while packing from nuspec

### PR DESCRIPTION
**Pull request by rohit21agrawal**
Fixes: https://github.com/NuGet/Home/issues/4250

This introduces new msbuild properties- 

a) NuspecProperties : This is a semicolon separated list of property value pairs , for eg: /p:NuspecProperties=\"authors=rohit;id=abcd\"

b) NuspecBasePath : This sets the base path for the input nuspec

This changeset also removes the requirement of having an assets file if you are packing using a nuspec file, i.e, if /p:NuspecFile is not an empty string.

CC: @rrelyea @rrelyea @emgarten @mishra14 @nkolev92 @alpaix 

**Note : I will update this PR with tests once https://github.com/NuGet/NuGet.Client/pull/1116 is checked in**

### CODE REVIEW

**Comment by alpaix** [comment]
Is there a spec for nuspec properties syntax?

**Comment by rohit21agrawal** [comment]
@alpaix it's the same as what we have currently, but i will update it in the Pack spec too

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs [hunk 1]
```diff
@@ -575,5 +582,22 @@ private void PopulateProjectAndPackageReferences(PackageBuilder packageBuilder,
...
             }
         }
+
+        private static IDictionary<string, string> ParsePropertiesAsDictionary(string[] properties)
+        {
+            var dictionary = new Dictionary<string, string>();
+            foreach (var item in properties)
+            {
+                string[] pair = item.Split('=');
+                if (pair.Length != 2)
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs]: Parsing failure should raise warning/error visible to a user.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTask.cs [hunk 2]
```diff
@@ -79,9 +84,12 @@ public override bool Execute()
         {
             var request = GetRequest();
             var logic = PackTaskLogic;
-
+            PackageBuilder packageBuilder = null;
             var packArgs = logic.GetPackArgs(request);
-            var packageBuilder = logic.GetPackageBuilder(request);
+            if (string.IsNullOrEmpty(request.NuspecFile))
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTask.cs:91]: Should this throw here? Is null value for `packageBuilder` allowed downstream?

Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTask.cs:91]: yes, the logic here is that :

if a nuspec file has been specified, then we don't need to build a package builder here as packing from nuspec doesn't rely on msbuild inputs.

if a nuspec file hasn't been specified, then we create packagebuilder object here and use it downstream.

both are valid codepaths.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs (hunk 1 above)
Comment (alpaix) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs]: What about `desc="In my description I ought to use = and ; characters just for fun"`?
Is this supported?

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTask.cs (hunk 2 above)
Comment (alpaix) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTask.cs:91]: It's not quite trivial from just looking at the code. Could you add a comment explaining the logic here?
Also, looking forward to tests covering both code paths.

Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTask.cs:91]: yes, as i said in the description, i'll update the PR with tests once https://github.com/NuGet/NuGet.Client/pull/1116 is checked in

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs (hunk 1 above)
Comment (emgarten) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs]: there needs to be a way to escape ``=`` and ``;``

Comment (alpaix) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs]: Split by `=` and use first token as property name and all the rest portion coming after the first `=` will be the property value with `"` stripped out.

Parsing is fun as it always involves tons of edge cases 😜 .

Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs]: @alpaix @emgarten wrote the tests and made changes as per feedback. all edge cases should be handled now. One thing to note is that msbuild automatically strips out the double quotes so we don't need to do that ourselves

**Review by emgarten — APPROVED** [review]

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs [hunk 3]
```diff
@@ -575,5 +585,26 @@ private void PopulateProjectAndPackageReferences(PackageBuilder packageBuilder,
                 }
             }
         }
+
+        private static IDictionary<string, string> ParsePropertiesAsDictionary(string[] properties)
+        {
+            var dictionary = new Dictionary<string, string>();
```
Comment (alpaix) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs]: Should it be case insensitive?

**Review by alpaix — APPROVED** [review]
Looks ✨ !

File: src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs (hunk 3 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack.Library/PackTaskLogic.cs]: good catch..thanks! fixed
